python/ray/autoscaler/_private/fake_multi_node/node_provider.py

```diff
@@ -18,6 +28,177 @@
+
+    bootstrap_cfg_path_on_container = "/home/ray/ray_bootstrap_config.yaml"
+    bootstrap_key_path_on_container = "/home/ray/ray_bootstrap_key.pem"
+
+    resources = resources or {}
+
+    cmd_kwargs = dict(
+        CMD_ENV="",
+        num_cpus=num_cpus,
+        num_gpus=num_gpus,
+        resources=json.dumps(resources, indent=None),
+        volume_dir=volume_dir,
+        autoscaling_config=bootstrap_cfg_path_on_container)
+
+    # Set to "auto" to mount current autoscaler directory to nodes for dev
+    fake_cluster_dev_dir = os.environ.get("FAKE_CLUSTER_DEV", "")
+    if fake_cluster_dev_dir:
+        if fake_cluster_dev_dir == "auto":
+            local_ray_dir = os.path.dirname(ray.__file__)
+        else:
+            local_ray_dir = fake_cluster_dev_dir
+        os.environ["FAKE_CLUSTER_DEV"] = local_ray_dir
+        docker_ray_dir = "/home/ray/anaconda3/lib/python3.7/site-packages/ray"
+        node_spec["volumes"] += [
+            f"{local_ray_dir}/autoscaler:{docker_ray_dir}/autoscaler:ro",
+        ]
+
+        cmd_kwargs["CMD_ENV"] = f"FAKE_CLUSTER_DEV={fake_cluster_dev_dir} "
+
+    if head:
+        node_spec["command"] = DOCKER_HEAD_CMD.format(**cmd_kwargs)
+        # Expose ports so we can connect to the cluster from outside
+        node_spec["ports"] = [
+            f"{host_gcs_port}:{DEFAULT_PORT}",
+            f"{host_object_manager_port}:8076",
+            f"{host_client_port}:10001",
+        ]
+        # Mount status and config files for the head node
+        node_spec["volumes"] += [
+            f"{host_dir(node_state_path)}:{node_state_path}",
+            f"{host_dir(docker_status_path)}:{docker_status_path}",
+            f"{host_dir(docker_compose_path)}:{docker_compose_path}",
+            f"{host_dir(bootstrap_config_path)}:"
+            f"{bootstrap_cfg_path_on_container}",
+            f"{host_dir(private_key_path)}:{bootstrap_key_path_on_container}",
+        ]
+
+        # Create file if it does not exist on local filesystem
+        for filename in [
+                node_state_path, docker_status_path, bootstrap_config_path
+        ]:
+            if not os.path.exists(filename):
+                with open(filename, "wt") as f:
+                    f.write("{}")
+    else:
+        node_spec["command"] = DOCKER_WORKER_CMD.format(**cmd_kwargs)
+        node_spec["depends_on"] = [FAKE_HEAD_NODE_ID]
+
+    # Mount shared directories and ssh access keys
+    node_spec["volumes"] += [
+        f"{host_dir(mounted_cluster_dir)}:/cluster/shared",
+        f"{host_dir(mounted_node_dir)}:/cluster/node",
+        f"{host_dir(public_key_path)}:/home/ray/.ssh/authorized_keys",
+    ]
+
+    env_vars = env_vars or {}
+
+    # Pass these environment variables (to the head node)
+    env_vars.setdefault("RAY_TEMPDIR", os.environ.get("RAY_TEMPDIR", ""))
+    env_vars.setdefault("RAY_HOSTDIR", os.environ.get("RAY_HOSTDIR", ""))
+
+    node_spec["environment"] = [f"{k}={v}" for k, v in env_vars.items()]
```

Comment:
where does this get consumed?

Reply:
This is saved in the `docker-compose.yaml` and consumed by the `docker compose` command (and propagated to the nodes). I'll add a comment for this.